### ISSUE TITLE
[runtime env] Can we avoid merging two runtime envs?

### ISSUE THREAD

**Issue by SongGuyang** [ray-project/ray#21494]
### Search before asking

- [X] I had searched in the [issues](https://github.com/ray-project/ray/issues) and found no similar feature requirement.

### Description

In current design, the actor level runtime env can inherit job level runtime env, like:
If the job level runtime env is:
```
{
   "py_modules": "s3://ab.zip",
    "pip": ["requests", "pendulum==2.1.2"],
    "env_vars":{
        "a":"b",
    }
}
```

and if the actor level runtime env is:
```
{
   "py_modules": "s3://cd.zip",
    "env_vars":{
        "c":"d",
    }
}
```

The final merged runtime env for actor is:
```
{
   "py_modules": "s3://cd.zip",
    "pip": ["requests", "pendulum==2.1.2"],
    "env_vars":{
        "a":"b",
        "c":"d",
    }
}
```

More details we can see https://github.com/ray-project/ray/blob/master/src/ray/core_worker/core_worker.cc#L1355 and https://github.com/ray-project/ray/blob/master/src/ray/core_worker/test/core_worker_test.cc#L937.

Do we have some reasons to support this logic? 
I think this logic is not clear and brings some complexity. Can we make it easier? Like：
- If there is no runtime env option when we create actor, inherit the job level runtime env.
- Otherwise, use the actor runtime env directly and don't do the merging.
- We also can give a new API to help users getting job level runtime env if needed.

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

**Comment by SongGuyang** [comment]
@edoakes @architkulkarni

**Comment by SongGuyang** [comment]
cc @raulchen

**Comment by raulchen** [comment]
I agree that merging runtime envs is hard to implement, and also could bring confusion to users. I'm in favor of the following logic. 

> If there is no runtime env option when we create actor, inherit the job level runtime env.
> Otherwise, use the actor runtime env directly and don't do the merging.
> We also can give a new API to help users getting job level runtime env if needed.

**Comment by architkulkarni** [comment]
I agree the current merging logic is a bit complicated--right now there is one special case (env_vars), and apart from that, each individual field gets overwritten with the child's field (like a simple `dict` merge).  But from the user's perspective, I'm not sure if the current approach or the proposed approach is more confusing.

Example 1: When a user specifies a new environment variable in a nested actor or task, they might expect the existing environment variables to be inherited (similar to how to how it works for processes and child processes.)

Example 2: If a job runtime env only specifies certain fields, like `"pip"`, then within the job the user might want to add an environment variable or a remote working_dir, just for a specific actor.  I think the user would expect the `pip` packages importable in the driver to continue to be importable inside the actor.  That's how it would work if the user didn't specify the per-actor runtime_env, and that's also how it would work if the user didn't use runtime_env at all.

**Comment by SongGuyang** [comment]
> Example 1: When a user specifies a new environment variable in a nested actor or task, they might expect the existing environment variables to be inherited (similar to how to how it works for processes and child processes.)

What we don't set all of the environment variable in the parent actor?

**Comment by SongGuyang** [comment]
> Example 2: If a job runtime env only specifies certain fields, like `"pip"`, then within the job the user might want to add an environment variable or a remote working_dir, just for a specific actor. I think the user would expect the `pip` packages to continue to be importable inside the actor. That's how it would work if the user didn't specify the per-actor runtime_env, and that's also how it would work if the user didn't use runtime_env at all.

What's the reason of using different working_dir for a specific actor? I think one global working dir is enough for most of the cases.

**Comment by SongGuyang** [comment]
Another case is that:
If you set a special conda env in parent actor, and set a special pip env in child actor, what's the final state? Conda env and pip env are conflicting.

Can we add a new API like `get_current_runtime_env` to support the special cases? We can open the merging logic to users to support some complex scenes.

**Comment by architkulkarni** [comment]
I don't have a concrete example for `env_var` question besides the analogy to processes, maybe @edoakes you know some other example?

For setting working_dir per-actor, one use case is for Ray Serve, where in a single Ray job there's many actors with independent ML model deployments, each one might need its own working_dir with its own data.  But maybe this is a special case where we should just use `get_current_runtime_env`.

For the third case (conda/pip), that's a really good point...  Right now we forbid "conda" and "pip" to both be specified in any runtime_env.  It seems like the user would have to manually specify `{"conda": None, pip: "new_pip_env.txt"}` in the child to get rid of the parent's `conda` field, but that's not a good experience and I'm not sure if it even works with the current merging logic.

Let me get some feedback from users on Ray Slack about the proposed change.

**Comment by rkooo567** [comment]
I also prefer to follow the proposed logic. It seems to be unstraightforward some values are inherited when runtime envs are given to actors (if you look at code, it is very implicit). Maybe we can alternatively do 

```
Actor.options(runtime_env=get_parent_runtime_env.update({"X": "Y"}))
```
? This seems a lot more explicit and less magic for user perspective.

**Comment by SongGuyang** [comment]
I think we have reached an agreement of this! I will create some PRs for this proposal.

**Comment by valiantljk** [comment]
> I also prefer to follow the proposed logic. It seems to be unstraightforward some values are inherited when runtime envs are given to actors (with the current style, it is hard to know what is the exact runtime env when reading code). Maybe we can alternatively do
> 
> ```
> Actor.options(runtime_env=get_parent_runtime_env.update({"X": "Y"}))
> ```
> 
> ? This seems a lot more explicit and less magic for user perspective.

I like this option, some users may assume job level env are passed to actor/task level, so previous merging doesn't always sound bad to me.

**Comment by edoakes** [comment]
@SongGuyang do you mind summarizing the proposal in an RFC issue so we can solicit more feedback?

**Comment by SongGuyang** [comment]
> @SongGuyang do you mind summarizing the proposal in an RFC issue so we can solicit more feedback?

Yep, if you think it is necessary, I will create a RFC issue. (I already have started to change code for this issue because it has an impact on subsequent PR. )

**Comment by edoakes** [comment]
Yeah the internal changes are fine, but before we make API changes such as adding the Python class for runtime_env I'd like to solicit more feedback

**Comment by architkulkarni** [comment]
We did solicit feedback in #general on the OSS slack on Tuesday: https://ray-distributed.slack.com/archives/C01DLHZHRBJ/p1642533302098800. It led to a couple comments on this issue.  So we might not need to make an RFC, but it's possible making a formal RFC will lead to even more comments.  I'll defer to @edoakes here.

**Comment by architkulkarni** [comment]
Ah I see, we only solicited feedback for this internal change (merging behavior) but we didn't solicit feedback for the API change (adding a Python class for runtime_env).  So it makes sense to have an RFC for the API change

**Comment by SongGuyang** [comment]
Make sense. I will create the RFC!

**Comment by SongGuyang** [comment]
> but we didn't solicit feedback for the API change (adding a Python class for runtime_env). 

Maybe we will only add a API like `get_current_runtime_env `, and current issue is not related to `RuntimeEnv` class. And I found we already has the API `ray.get_runtime_context().runtime_env`. Maybe we have no need to add a new API for current issue.

**Comment by SongGuyang** [comment]
But if we do that, we will change the behavior of current runtime env API. So, it makes sense to create the RRC.

**Comment by rkooo567** [comment]
Is runtime env GA or beta? I think we need to be careful changing the default behavior if it is GA already

**Comment by SongGuyang** [comment]
Seems already GA in 1.6.0 https://github.com/ray-project/ray/releases/tag/ray-1.6.0.

**Comment by cszhu** [comment]
Hello! This P2 issue has seen no activity in at least 2 years. It will be closed in 2 weeks as part of ongoing cleanup efforts.

Please remove the `pending-cleanup` label if you believe this issue should remain open.

Thanks for contributing to Ray!
